runtime_scan/pkg/orchestrator/discovery/scope.go

```diff
@@ -65,3 +64,74 @@ func (sd *ScopeDiscoverer) Start(ctx context.Context) {
 		}
 	}()
 }
+
+func (sd *ScopeDiscoverer) DiscoverAndCreateAssets(ctx context.Context) error {
+	discoveryTime := time.Now()
+
+	assetTypes, err := sd.providerClient.DiscoverTargets(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to discover assets from provider: %w", err)
+	}
+
+	errs := []error{}
+	for _, assetType := range assetTypes {
+		targetData := models.Target{
+			TargetInfo: utils.PointerTo(assetType),
+			LastSeen:   &discoveryTime,
+		}
+		_, err := sd.backendClient.PostTarget(ctx, targetData)
+		if err == nil {
+			continue
+		}
+
+		var conflictError backendclient.TargetConflictError
+		if !errors.As(err, &conflictError) {
+			// If there is an error, and its not a conflict telling
+			// us that the target already exists, then we need to
+			// keept track of it and log it as a failure to
+			// complete discovery. We don't fail instantly here
+			// because discovering the assets is a heavy operation
+			// so we want to give the best chance to create all the
+			// assets in the DB before failing.
+			errs = append(errs, fmt.Errorf("failed to post target: %v", err))
+			continue
+		}
+
+		// As we got a conflict it means there is an existing target
+		// which matches the unique properties of this target, in this
+		// case we'll patch the data instead.
+		err = sd.backendClient.PatchTarget(ctx, targetData, *conflictError.ConflictingTarget.Id)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to patch target: %v", err))
+		}
+	}
+
+	// Find all assets which were not updated or created by this discovery
+	// run by comparing their lastSeen time to this discovery's time stamp
+	//
+	// TODO(sambetts) when we add multiple providers/standalone support we
+	// need to filter these assets by provider so that we don't find assets
+	// which don't belong to us. We need to give the provider some kind of
+	// identity in this case.
+	targetResp, err := sd.backendClient.GetTargets(ctx, models.GetTargetsParams{
+		Filter: utils.PointerTo(fmt.Sprintf("lastSeen eq null or lastSeen lt %s", discoveryTime.Format(time.RFC3339))),
```

Comment:
Added not terminated to the query. Good point on marking it terminated if we fail to patch, perhaps I should skip the terminated processing if there are errors during the first part because we don't know if we've got completely accurate data?

The other thing I was thinking was to have the discovery be able to unset terminated, so if an asset disappears, but then reappears later, we can unset terminated effectively re-instating the asset. In that case if there was an error patching it, it would get marked terminated, but then the next discovery would un-terminate it?

Reply:
https://github.com/openclarity/vmclarity/issues/446